rllib/models/base_model.py

```diff
@@ -0,0 +1,330 @@
+                timestep of the previous unroll.
+            kwargs: For forwards compatibility
+
+        Returns:
+            outputs: A TensorDict of outputs
+            next_state: A dict containing the state to be passed
+                as the first state of the next rollout.
+        """
+
+    @DeveloperAPI
+    def unroll(
+        self, inputs: TensorDict, prev_state: TensorDict, **kwargs
+    ) -> UnrollOutputType:
+        """Computes the output of the module over unroll_len timesteps.
+
+        Args:
+            inputs: A TensorDict containing inputs to the model
+            prev_state: A TensorDict containing containing the
+                next_state of the last timestep of the previous unroll.
+            kwargs: For forwards compatibility
+
+        Returns:
+            outputs: A TensorDict containing model outputs
+            next_state: A TensorDict containing the
+                state to be passed as the first state of the next rollout.
+
+        Examples:
+            >>> output, state = model.unroll(TensorDict(...), TensorDict(...))
+            >>> output # TensorDict(...)
+            >>> state # TensorDict(...)
+
+        """
+        self.input_spec.validate(inputs)
+        self.prev_state_spec.validate(prev_state)
+        # We hide inputs not specified in input_spec to prevent accidental use.
+        inputs = inputs.filter(self.input_spec)
+        prev_state = prev_state.filter(self.prev_state_spec)
+        inputs, prev_state = self._check_inputs_and_prev_state(inputs, prev_state)
+        outputs, next_state = self._unroll(inputs, prev_state, **kwargs)
+        self.output_spec.validate(outputs)
+        self.next_state_spec.validate(next_state)
+        outputs, next_state = self._check_outputs_and_next_state(outputs, next_state)
+        return outputs, next_state
+
+    def _check_inputs_and_prev_state(
+        self, inputs: TensorDict, prev_state: TensorDict
+    ) -> Tuple[TensorDict, TensorDict]:
+        """Override this function to add additional checks on inputs.
+
+        Args:
+            inputs: TensorDict containing inputs to the model
+            prev_state: The previous recurrent state
+
+        Returns:
+            inputs: Potentially modified inputs
+            prev_state: Potentially modified recurrent state
+        """
+        return inputs, prev_state
+
+    def _check_outputs_and_next_state(
+        self, outputs: TensorDict, next_state: TensorDict
+    ) -> Tuple[TensorDict, TensorDict]:
+        """Override this function to add additional checks on outputs.
+
+        Args:
+            outputs: TensorDict output by the model
+            next_state: Recurrent state output by the model
+
+        Returns:
+            outputs: Potentially modified TensorDict output by the model
+            next_state: Potentially modified recurrent state output by the model
+        """
+        return outputs, next_state
+
+
+class Model(RecurrentModel):
```

Comment:
so Model is a special case of RecurrentModel?
why not RecurrentModel is Model + input/output states?
this may seem weird.

Reply:
@gjoliver Think about it from the point of view of which one is more general to serve as the base class? Every non-recurrent model can be expressed as a Recurrent Model but the opposite is not true.

Reply:
dude, I know we just had a call about this. but if you think about it, the reverse has to be True, like at the end of the day, RNN models need to be reduced to a sequence of matrix multiplications, which is model.
I definitely don't want to muddy the water, and get us in any philosophical debate of NNs, but is there any chance our Model definition is not low level enough? For example, instead of ``unroll()``, should we have something like ``__call__(Tuple) -> Tuple``?
I promise this is the last message I will send about this topic.

Reply:
@gjoliver how about we merge this abstract class for now and in the next phase where we actually use this model to build some default base components, revisit this design choices and modify things there? Without the next stage's concrete examples this discussion won't get to anywhere.

Reply:
fine, fine